In [40]:
import pandas as pd 

In [44]:
# Load the dataset and check

rd_budget_df = pd.read_csv('data/rd_data.csv')
rd_budget_df.head()
print(rd_budget_df)

   department  1976_gdp1790000000000.0  1977_gdp2028000000000.0  \
0         DHS                      NaN                      NaN   
1         DOC             8.190000e+08             8.370000e+08   
2         DOD             3.569600e+10             3.796700e+10   
3         DOE             1.088200e+10             1.374100e+10   
4         DOT             1.142000e+09             1.095000e+09   
5         EPA             9.680000e+08             9.660000e+08   
6         HHS             9.226000e+09             9.507000e+09   
7    Interior             1.152000e+09             1.082000e+09   
8        NASA             1.251300e+10             1.255300e+10   
9         NIH             8.025000e+09             8.214000e+09   
10        NSF             2.372000e+09             2.395000e+09   
11      Other             1.191000e+09             1.280000e+09   
12       USDA             1.837000e+09             1.796000e+09   
13         VA             4.040000e+08             3.740000e+0

The dataset initially has 14 rows and 43 columns, with rows indicating each department and columns indicating each year and corresposnding budget. This layout is not convenient to read so we need to reshape the format of the dataset. 

In [45]:
# Reshape the dataset from a wide format to a long format 
rd_budget_long = rd_budget_df.melt(id_vars=["department"], var_name="year_gdp", value_name="budget")
print(rd_budget_long)

    department                  year_gdp        budget
0          DHS   1976_gdp1790000000000.0           NaN
1          DOC   1976_gdp1790000000000.0  8.190000e+08
2          DOD   1976_gdp1790000000000.0  3.569600e+10
3          DOE   1976_gdp1790000000000.0  1.088200e+10
4          DOT   1976_gdp1790000000000.0  1.142000e+09
..         ...                       ...           ...
583        NIH  2017_gdp19177000000000.0  3.305200e+10
584        NSF  2017_gdp19177000000000.0  6.040000e+09
585      Other  2017_gdp19177000000000.0  1.553000e+09
586       USDA  2017_gdp19177000000000.0  2.625000e+09
587         VA  2017_gdp19177000000000.0  1.367000e+09

[588 rows x 3 columns]


This code reshapes the dataset using melt function that I found online. The column department remains as it is in the new format. The year columns are combined into a single column called year_gdp. The values in the year columns are put into a new column called year_gdp. 
Now the dataset has 588 rows and 3 columns. The first column represents the department, the second is the year and third is the budget itself. This is still not very convenient and clean so we need to proceed forther. 

In [46]:
# Extract the year from "year_gdp"
rd_budget_long["year"] = rd_budget_long["year_gdp"].str.split("_").str[0]

# Extract the GDP by removing the "gdp" prefix
rd_budget_long["gdp"] = rd_budget_long["year_gdp"].str.split("_").str[1].str.replace("gdp", "", regex=False)

# Drop the original "year_gdp" column
rd_budget_long = rd_budget_long.drop(columns=["year_gdp"])

# Convert "year" and "gdp" to numeric types
rd_budget_long["year"] = pd.to_numeric(rd_budget_long["year"], errors="coerce")
rd_budget_long["gdp"] = pd.to_numeric(rd_budget_long["gdp"], errors="coerce")

# Display result
print(rd_budget_long)

    department        budget  year           gdp
0          DHS           NaN  1976  1.790000e+12
1          DOC  8.190000e+08  1976  1.790000e+12
2          DOD  3.569600e+10  1976  1.790000e+12
3          DOE  1.088200e+10  1976  1.790000e+12
4          DOT  1.142000e+09  1976  1.790000e+12
..         ...           ...   ...           ...
583        NIH  3.305200e+10  2017  1.917700e+13
584        NSF  6.040000e+09  2017  1.917700e+13
585      Other  1.553000e+09  2017  1.917700e+13
586       USDA  2.625000e+09  2017  1.917700e+13
587         VA  1.367000e+09  2017  1.917700e+13

[588 rows x 4 columns]


Here I cleaned the year column to leave only the year using the split function by splitting the column before and after the lower _ and taking the first part of it with the index 0 that corresposnds to the year and created a new column called "year". To avoid dyplicates I deleted the intil column that contained original information using the methof drop and specifying the column. Tjen I converted the new colymn to integer type to be able to perform calculateions and any other functions later. 

I still dont necessaruly like the layout of the table and I want to transform it in the way that cyears will be columns and gdp budget will be values so that it will be easier to visualisr the data. I coud have dne this but its not a tidy data principle. Even though it mat be more convenient I will still use thge other format

In [35]:
rd_budget_messy = rd_budget_long.pivot(index="department", columns="year", values="budget")
rd_budget_messy.head()

year                1976          1977          1978          1979  \
department                                                           
DHS                  NaN           NaN           NaN           NaN   
DOC         8.190000e+08  8.370000e+08  8.710000e+08  9.520000e+08   
DOD         3.569600e+10  3.796700e+10  3.702200e+10  3.717400e+10   
DOE         1.088200e+10  1.374100e+10  1.566300e+10  1.561200e+10   
DOT         1.142000e+09  1.095000e+09  1.156000e+09  1.004000e+09   

year                1980          1981          1982          1983  \
department                                                           
DHS                  NaN           NaN           NaN           NaN   
DOC         9.450000e+08  8.290000e+08  6.930000e+08  7.000000e+08   
DOD         3.700500e+10  4.173700e+10  4.689300e+10  5.100800e+10   
DOE         1.522600e+10  1.479800e+10  1.216000e+10  1.112900e+10   
DOT         1.048000e+09  9.780000e+08  7.020000e+08  8.000000e+08   

year                1984          1985  ...          2008          2009  \
department                              ...                               
DHS                  NaN           NaN  ...  1.155000e+09  1.258000e+09   
DOC         7.550000e+08  7.890000e+08  ...  1.379000e+09  1.594000e+09   
DOD         5.813900e+10  6.441600e+10  ...  9.424700e+10  9.352700e+10   
DOE         1.173100e+10  1.204300e+10  ...  1.134300e+10  1.182300e+10   
DOT         1.018000e+09  8.850000e+08  ...  1.016000e+09  1.062000e+09   

year                2010          2011          2012          2013  \
department                                                           
DHS         1.009000e+09  8.480000e+08  5.270000e+08  7.370000e+08   
DOC         1.521000e+09  1.357000e+09  1.373000e+09  1.397000e+09   
DOD         9.432500e+10  8.822000e+10  8.154300e+10  7.058300e+10   
DOE         1.232900e+10  1.190200e+10  1.183900e+10  1.152800e+10   
DOT         1.221000e+09  1.064000e+09  1.007000e+09  8.810000e+08   

year                2014          2015          2016          2017  
department                                                          
DHS         1.092000e+09  9.600000e+08  6.020000e+08  7.350000e+08  
DOC         1.641000e+09  1.596000e+09  1.730000e+09  1.824000e+09  
DOD         7.034100e+10  6.952200e+10  7.521300e+10  5.186200e+10  
DOE         1.268600e+10  1.503300e+10  1.550400e+10  1.503200e+10  
DOT         8.430000e+08  9.260000e+08  9.100000e+08  9.360000e+08  

[5 rows x 42 columns]

I used pivot function to transform the table. This means that each row will represent a different department. This will turn the unique values in the year column into separate columns. The budget values will fill the cells in the table.

This setup looks more convenient and understandable. Now I need to check for missing values and do something with the, 

In [36]:
missing_per_year = rd_budget_long.isnull().sum()
print(missing_per_year)

department     0
budget        26
year           0
dtype: int64


In [37]:
missing_per_department = rd_budget_long.isnull().sum(axis=1)
print(missing_per_department)

0      1
1      0
2      0
3      0
4      0
      ..
583    0
584    0
585    0
586    0
587    0
Length: 588, dtype: int64


In [39]:
total_missing = rd_budget_long.isnull().sum().sum()
print("Total missing values:", total_missing)

Total missing values: 26


There are missing values only in the DHS department. The data for this department starts in 2002. So generally the data is clean but we will replace missing values with 0 for the purpose 

In [ ]:
rd_budget_long.fillna(0, inplace=True)
rd_budget_long.head()

year                1976          1977          1978          1979  \
department                                                           
DHS         0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
DOC         8.190000e+08  8.370000e+08  8.710000e+08  9.520000e+08   
DOD         3.569600e+10  3.796700e+10  3.702200e+10  3.717400e+10   
DOE         1.088200e+10  1.374100e+10  1.566300e+10  1.561200e+10   
DOT         1.142000e+09  1.095000e+09  1.156000e+09  1.004000e+09   

year                1980          1981          1982          1983  \
department                                                           
DHS         0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
DOC         9.450000e+08  8.290000e+08  6.930000e+08  7.000000e+08   
DOD         3.700500e+10  4.173700e+10  4.689300e+10  5.100800e+10   
DOE         1.522600e+10  1.479800e+10  1.216000e+10  1.112900e+10   
DOT         1.048000e+09  9.780000e+08  7.020000e+08  8.000000e+08   

year                1984          1985  ...          2008          2009  \
department                              ...                               
DHS         0.000000e+00  0.000000e+00  ...  1.155000e+09  1.258000e+09   
DOC         7.550000e+08  7.890000e+08  ...  1.379000e+09  1.594000e+09   
DOD         5.813900e+10  6.441600e+10  ...  9.424700e+10  9.352700e+10   
DOE         1.173100e+10  1.204300e+10  ...  1.134300e+10  1.182300e+10   
DOT         1.018000e+09  8.850000e+08  ...  1.016000e+09  1.062000e+09   

year                2010          2011          2012          2013  \
department                                                           
DHS         1.009000e+09  8.480000e+08  5.270000e+08  7.370000e+08   
DOC         1.521000e+09  1.357000e+09  1.373000e+09  1.397000e+09   
DOD         9.432500e+10  8.822000e+10  8.154300e+10  7.058300e+10   
DOE         1.232900e+10  1.190200e+10  1.183900e+10  1.152800e+10   
DOT         1.221000e+09  1.064000e+09  1.007000e+09  8.810000e+08   

year                2014          2015          2016          2017  
department                                                          
DHS         1.092000e+09  9.600000e+08  6.020000e+08  7.350000e+08  
DOC         1.641000e+09  1.596000e+09  1.730000e+09  1.824000e+09  
DOD         7.034100e+10  6.952200e+10  7.521300e+10  5.186200e+10  
DOE         1.268600e+10  1.503300e+10  1.550400e+10  1.503200e+10  
DOT         8.430000e+08  9.260000e+08  9.100000e+08  9.360000e+08  

[5 rows x 42 columns]

The final thing left to do is to make numbers more readable and format them in a more human friendly way instead of a scientific notation. 